_posts/2020-09-29-introducing-torch-for-r/introducing_torch_for_R.Rmd

```diff
@@ -0,0 +1,417 @@
+```
+
+Again, `torch` makes it easy to verify we did the correct thing. To take a look at the content of the first batch, do
+
+```{r}
+train_iter <- train_dl$.iter()
+train_iter$.next()
+```
+
+Functionality like this may not seem indispensable when working with a well-known dataset, but will turn out to be very useful when a lot of domain-specific pre-processing is required.
+
+Now that we've seen how to load data, all prerequisites are fulfilled for visualizing them. Here is the code that was used to display the first batch of characters, above:
+
+```{r, fig.asp=0.5, fig.width=8}
+par(mfrow = c(4,8), mar = rep(0, 4))
+images <- train_dl$.iter()$.next()[[1]][1:32, 1, , ] 
+images %>%
+  purrr::array_tree(1) %>%
+  purrr::map(as.raster) %>%
+  purrr::iwalk(~{plot(.x)})
+```
+
+We're ready to define our network -- a simple convnet.
+
+## Network
+
+If you've been using `keras` *custom models* (or have some experience with *Py*Torch), the following way of defining a network may not look too surprising.
+
+You use `nn_module()` to define an R6 class that will hold the network's components. Its layers are created in `initialize()`; `forward()` describes what happens during the network's forward pass. One thing on terminology: In `torch`, layers are called *modules*, as are networks. This makes sense: The design is truly *modular* in that any module can be used as a component in a larger one.
+
+```{r}
+net <- nn_module(
+  
+  "KMNIST-CNN",
+  
+  initialize = function() {
+    # in_channels, out_channels, kernel_size, stride = 1, padding = 0
+    self$conv1 <- nn_conv2d(1, 32, 3)
+    self$conv2 <- nn_conv2d(32, 64, 3)
+    self$dropout1 <- nn_dropout2d(0.25)
+    self$dropout2 <- nn_dropout2d(0.5)
+    self$fc1 <- nn_linear(9216, 128)
+    self$fc2 <- nn_linear(128, 10)
+  },
+  
+  forward = function(x) {
+    x %>% 
+      self$conv1() %>%
+      nnf_relu() %>%
+      self$conv2() %>%
+      nnf_relu() %>%
+      nnf_max_pool2d(2) %>%
+      self$dropout1() %>%
+      torch_flatten(start_dim = 2) %>%
+      self$fc1() %>%
+      nnf_relu() %>%
+      self$dropout2() %>%
+      self$fc2()
+  }
+)
+```
+
+The layers -- apologies: modules -- themselves may look familiar. Unsurprisingly, `nn_conv2d()` performs two-dimensional convolution; `nn_linear()` multiplies by a weight matrix and adds a vector of biases. But what are those numbers: `nn_linear(128, 10)`, say?
+
+In `torch`, instead of the number of units in a layer, you specify input and output dimensionalities of the "data" that run through it. Thus, `nn_linear(128, 10)` has 128 input connections and outputs 10 values -- one for every class. In some cases, such as this one, specifying dimensions is easy -- we know how many input edges there are (namely, the same as the number of output edges from the previous layer), and we know how many output values we need. But how about this module's predecessor? How do we arrive at `9216` input connections?
```

Comment:
I'm a little confused about this:
> But how about this module's predecessor? How do we arrive at `9216` input connections?

OK, I just un-confused myself by scrolling back up in the code, and I see that you're referring to:
```r
    self$fc1 <- nn_linear(9216, 128)
    self$fc2 <- nn_linear(128, 10)
```

I think it might be easier to follow if you reference them as `fc1` and `fc2` along with giving the dimensions. Alone I wasn't totally sure what you meant by "predecessor"—I guess I was thinking of it as the preceding step.

Reply:
i think i should change to "preceding module"